File: java/leetcode/InorderTraversal.java
//Leetcode problem 94 Binary Tree Inorder Traversal
//Solution written by Xuqiang Fang on 13 March 2018 
//Iterative solution provided by user@lvlolitte
import java.util.ArrayList;
import java.util.List;
import java.util.LinkedList;


class TreeNode{
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x){val = x;}
}
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
		List<Integer> list = new ArrayList<Integer>();
		LinkedList<TreeNode> stack = new LinkedList<TreeNode>();
		TreeNode current = root;
		while(current != null || !stack.isEmpty()){
			while(current != null){
				stack.push(current);
				current = current.left;
			}

			current = stack.pop();
			list.add(current.val);
			current = current.right;
		}
		return list;
    }
}


public class InorderTraversal{
	public static void main(String[] args){
		Solution s = new Solution();
		TreeNode root = new TreeNode(1);
		TreeNode a = new TreeNode(2);
		TreeNode b = new TreeNode(3);
		root.right = a;
		a.left = b;
		System.out.println(s.inorderTraversal(root));
	}
}

File: python/leetcode/InorderTraversal.py
'''
Leetcode problem No 94 Binary Tree Inorder Traversal
Solution written by Xuqiang Fang on 20 July, 2018
'''

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        ans, stack = [], [(root, False)]
        while stack:
            root, visited = stack.pop()
            if not root:
                continue
            if visited:
                ans.append(root.val)
            else:
                stack.append((root.right, False))
                stack.append((root, True))
                stack.append((root.left, False))
        return ans

def main():
    s = Solution()
    r = TreeNode(1)
    a = TreeNode(2)
    b = TreeNode(3)
    c = TreeNode(4)
    d = TreeNode(5)
    r.left = a
    r.right = b
    a.left = c
    a.right = d
    ans = s.inorderTraversal(r)
    for i in ans:
        print(i)
if __name__ == '__main__':
    main()
